pkg/agent/externalnode/external_node_controller_test.go

```diff
@@ -0,0 +1,573 @@
+	mockOVSBridgeClient.EXPECT().DeletePort(uplinkUUID).Times(1)
+	mockOVSBridgeClient.EXPECT().DeletePort(hostIfUUID).Times(1)
+	mockOFClient.EXPECT().InstallVMUplinkFlows(hostIfName, hostOFPort, uplinkOFPort).Times(1)
+	hostIFConfig, err := c.createOVSPortsAndFlows(uplinkName, hostIfName, entityNamespace, entityName, ipAddrs)
+	assert.Equal(t, ovsconfig.NewTransactionError(fmt.Errorf("interface %s not found", hostIfName), false), err)
+	assert.Nil(t, hostIFConfig)
+}
+
+func TestUpdateOVSPortsData(t *testing.T) {
+	c := newExternalNodeController(t)
+	interfaceConfig := &intf1
+	eeName := "eeName"
+	ips := []string{"10.20.30.40"}
+	attachInfo := map[string]interface{}{
+		"antrea-type":      "host",
+		"entity-name":      "eeName",
+		"entity-namespace": "externalEntityNamespace",
+		"ip-address":       "10.20.30.40",
+		"uplink-name":      "uplinkName",
+		"uplink-port":      portUUID2,
+	}
+	mockOVSBridgeClient.EXPECT().SetPortExternalIDs(intf1.InterfaceName, attachInfo)
+	expectedIface := &interfacestore.InterfaceConfig{
+		InterfaceName: interfaceConfig.InterfaceName,
+		Type:          interfacestore.ExternalEntityInterface,
+		OVSPortConfig: &interfacestore.OVSPortConfig{
+			PortUUID: interfaceConfig.PortUUID,
+			OFPort:   interfaceConfig.OFPort,
+		},
+		EntityInterfaceConfig: &interfacestore.EntityInterfaceConfig{
+			EntityName:      eeName,
+			EntityNamespace: interfaceConfig.EntityNamespace,
+			UplinkPort: &interfacestore.OVSPortConfig{
+				PortUUID: interfaceConfig.UplinkPort.PortUUID,
+				OFPort:   interfaceConfig.UplinkPort.OFPort,
+			},
+		},
+		IPs: []net.IP{net.ParseIP("10.20.30.40")},
+	}
+	iface, err := c.updateOVSPortsData(interfaceConfig, portData1, eeName, ips)
+	assert.Equal(t, expectedIface, iface)
+	assert.NoError(t, err)
+}
+
+func TestParseProtocol(t *testing.T) {
+	for _, tt := range []struct {
+		name           string
+		protocol       string
+		expectProtocol binding.Protocol
+	}{
+		{
+			name:           "UDP protocol",
+			protocol:       "udp",
+			expectProtocol: binding.ProtocolUDP,
+		},
+		{
+			name:           "TCP protocol",
+			protocol:       "tcp",
+			expectProtocol: binding.ProtocolTCP,
+		},
+		{
+			name:           "ICMP protocol",
+			protocol:       "icmp",
+			expectProtocol: binding.ProtocolICMP,
+		},
+		{
+			name:           "IP protocol",
+			protocol:       "ip",
+			expectProtocol: binding.ProtocolIP,
+		},
+		{
+			name:           "IGMP protocol",
+			protocol:       "igmp",
+			expectProtocol: "",
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			protocol := parseProtocol(tt.protocol)
+			assert.Equal(t, tt.expectProtocol, protocol)
+		})
+	}
+}
+
+func TestParseHostInterfaceConfig(t *testing.T) {
+	controller := mock.NewController(t)
+	defer controller.Finish()
+
+	randomUUID := uuid.NewString()
+	portUUID := uuid.NewString()
+	mockOVSBridgeClient := ovsconfigtest.NewMockOVSBridgeClient(controller)
```

Comment:
move `controller` and `mockOVSBridgeClient` to `t.Run()` to avoid side effect: https://github.com/tnqn/code-review-comments#table-driven-test

Reply:
updated